train_model_lightning.py

```diff
@@ -0,0 +1,188 @@
+
+    # Initialize checkpoints directory
+    checkpoints_dir = model_dir / "checkpoints"
+    checkpoints_dir.mkdir(parents=False, exist_ok=False)
+    print(f"Saving checkpoints in {checkpoints_dir}")
+
+    # Create SummaryWriter to record logs for TensorBoard
+    if config.tboard_path is None:
+        tboard_log_dir = Path(config.models_path) / "logs" / model_label
+    else:
+        tboard_log_dir = f"{config.tboard_path}/logs/{model_label}"
+
+    print(f"Saving TensorBoard logs in {tboard_log_dir}")
+
+    # Save all the variables in args as JSON inside folder
+    json.dump(
+        obj=arg_table,
+        fp=open(model_dir / "model_args.json", "w"),
+        indent=4)
+
+    # Print estimated loss if it hasn't learned anything
+    print("\nEstimated Loss if guessing:")
+    print(f"-log(1 / {config.vocab_size}) = " + \
+        f"{-torch.log(torch.tensor(1 / config.vocab_size))}")
+
+    # Loads Tokenized data
+    print(f"\nNow loading '{config.dataset_name}'")
+
+    dm = DataModule(config, num_workers=1)
+    
+    # Implement callbacks
+    model_checkpoint = CustomCheckpoint(
+        dirpath=checkpoints_dir,
+        filename='epoch_{epoch}_validation_{val_loss:.2f}',
+        save_top_k=config.save_top_k,
+        monitor='val_loss',
+        mode='min')
+    
+    if not config.use_slurm:
+        trainer = Trainer(
+            default_root_dir=model_dir, # main directory for run
+            accelerator=config.device,
+            devices=config.num_devices,
+            max_epochs=config.epochs,
+            val_check_interval=config.val_check_interval,
+            accumulate_grad_batches=config.accumulate_grad_batches,
+            sync_batchnorm=True,
+            callbacks=[model_checkpoint]
+            )
+    else:
+        trainer = Trainer(
+            default_root_dir=model_dir, # main directory for run
+            accelerator=config.device,
+            num_nodes=config.num_nodes,
+            devices=config.num_devices,
+            strategy=config.strategy,
+            max_epochs=config.epochs,
+            val_check_interval=config.val_check_interval,
+            accumulate_grad_batches=config.accumulate_grad_batches,
+            sync_batchnorm=True,
+            plugins=[SLURMEnvironment(requeue_signal=signal.SIGHUP)],
+            callbacks=[model_checkpoint]
+            )
+    
+    trainer.fit(model, datamodule=dm)
+
+    print("\nDone training! Now testing model...")
+
+    trainer.test(datamodule=dm) # Automatically loads best checkpoint, and tests with test dataloader
+
+    print('Finished training!')
+
+    return model, trainer.callback_metrics['val_loss'].item()
+
+if __name__ == "__main__":
+    args = sys.argv
+    config_path =args[1]
+
+    with open(config_path, 'r') as f:
+        config = yaml.safe_load(f)
+        
+    config = Struct(**config)
+
+    start_time = time.time()
+
+    model, test_loss = train_model(config)
+
+    end_time = time.time()
```

Comment:
What is the point of these times? It doesn't seem that they're used anywhere....